python/katana/local/_graph.pyx

```diff
@@ -229,15 +229,16 @@ cdef class GraphBase:
         """
         handle_result_void(self.underlying_property_graph().AddNodeProperties(GraphBase._convert_table(table, kwargs)))
 
-    def upsert_node_property(self, table=None, **kwargs):
+    def upsert_node_property(self, PyTxnContext ctx, table=None, **kwargs):
```

Comment:
Is there a safe default value for `ctx`? We should support non-transactional calls for simple uses.

Reply:
In the long run I think we should track the transaction in a thread local at the python level. It's much more natural for the programmer to write `with transaction(): ...` than to have to explicitly manage the context object. This shouldn't happen in the PR. But we should make sure we have a reasonable path to it.

Reply:
The `TxnContext` right now only stores read and write set information. The ctx passed into the entry point contains empty read and write sets. If the users does not need transactions, they can simply ignore the content of ctx. But I think we assume single operations outside transactions to be single transaction operations as well. 

I agree on the thread local python transaction. But do you think it'll affect c++ implementation?

Reply:
I don't think it will affect C++. Python can keep the object around as needed without needing to involve C++. As long as C++ doesn't make assumptions about the threading model.

Reply:
I added default value for `TxnContext` and put the `track the transaction in a thread local at the python level` as a Jira task KAT-2568 but don't know how to assign you as reporter. It says reporter cannot be changed.